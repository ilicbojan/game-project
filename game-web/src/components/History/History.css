.history ul {
  list-style: none;
  padding: 0;
}

.history li {
  padding: 0.5rem 0;
  border-bottom: 1px solid #eee;
}

.history li:last-child {
  border-bottom: none;
}

.history-player,
.history-computer,
.history-result {
  font-weight: bold;
}
