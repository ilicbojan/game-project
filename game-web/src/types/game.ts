export interface ChoiceItem {
  id: number;
  name: string;
}

export interface PlayRequest {
  player: number;
}

export interface PlayResponse {
  player: number;
  computer: number;
  result: ResultType;
}

export interface GameResult {
  player: string;
  computer: string;
  result: ResultType;
  date: Date;
}

export interface Score {
  win: number;
  lose: number;
  tie: number;
}

export enum ResultType {
  Win = "win",
  Lose = "lose",
  Tie = "tie",
}

export enum ChoiceId {
  Rock = 1,
  Paper = 2,
  Scissors = 3,
  Lizard = 4,
  Spock = 5,
}

export enum ChoiceName {
  Rock = "rock",
  Paper = "paper",
  Scissors = "scissors",
  Lizard = "lizard",
  Spock = "spock",
}
