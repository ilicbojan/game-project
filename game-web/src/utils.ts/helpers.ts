import { ChoiceId, ChoiceItem, ChoiceName, ResultType } from "../types/game";

export const ResultToColorMap: Record<ResultType, string> = {
  [ResultType.Win]: "green",
  [ResultType.Lose]: "red",
  [ResultType.Tie]: "orange",
};

export const ChoiceToEmojiMap: Record<string, string> = {
  rock: "✊", // Rock
  paper: "✋", // Paper
  scissors: "✌️", // Scissors
  lizard: "🦎", // Lizard
  spock: "🖖", // Spock
  random: "🎲", // Random choice
  "": "❓", // Unknown choice
};

const choiceItemMap: Record<string, ChoiceItem> = {
  rock: { id: ChoiceId.Rock, name: ChoiceName.Rock },
  paper: { id: ChoiceId.Paper, name: ChoiceName.Paper },
  scissors: { id: ChoiceId.Scissors, name: ChoiceName.Scissors },
  lizard: { id: ChoiceId.Lizard, name: ChoiceName.Lizard },
  spock: { id: ChoiceId.Spock, name: ChoiceName.Spock },
};

export function getChoiceItem(name: ChoiceName): ChoiceItem {
  return choiceItemMap[name];
}

export const formatDateTime = (d: Date) =>
  `${d.toLocaleDateString()} ${d.toLocaleTimeString()}`;
