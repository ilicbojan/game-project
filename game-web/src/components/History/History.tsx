import React from "react";
import { GameResult } from "../../types/game";
import "./History.css";
import { ChoiceToEmojiMap, ResultToColorMap } from "../../utils.ts/helpers";

interface Props {
  history: GameResult[];
}

const History: React.FC<Props> = ({ history }) => {
  return (
    <div className="history section">
      <h3>
        History <span>{history.length} / 10</span>
      </h3>
      <ul>
        {history.slice(0, 10).map((item, index) => (
          <li key={index}>
            <span className="history-date">
              {item.date.toLocaleDateString()} {item.date.toLocaleTimeString()}{" "}
              |{" "}
            </span>
            You chose{" "}
            <span className="history-player">
              {item.player} {ChoiceToEmojiMap[item.player]}
            </span>
            , computer chose{" "}
            <span className="history-computer">
              {item.computer} {ChoiceToEmojiMap[item.computer]}
            </span>{" "}
            â†’{" "}
            <span
              className="history-result"
              style={{ color: ResultToColorMap[item.result] }}
            >
              {item.result}
            </span>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default History;
