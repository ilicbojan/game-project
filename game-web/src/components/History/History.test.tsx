import { render, screen } from "@testing-library/react";
import { ChoiceName, GameResult, ResultType } from "../../types/game";
import History from "./History";
import { ResultToColorMap } from "../../utils.ts/helpers";

const dateStringMock = "1/1/2025 10:00:00 AM";
jest.mock("../../utils.ts/helpers", () => ({
  ...jest.requireActual("../../utils.ts/helpers"),
  formatDateTime: () => dateStringMock,
}));

describe("History component", () => {
  const dateMock = new Date(dateStringMock);
  const historyMock: GameResult[] = [
    {
      player: ChoiceName.Rock,
      computer: ChoiceName.Scissors,
      result: ResultType.Win,
      date: dateMock,
    },
    {
      player: ChoiceName.Paper,
      computer: ChoiceName.Spock,
      result: ResultType.Win,
      date: dateMock,
    },
    {
      player: ChoiceName.Lizard,
      computer: ChoiceName.Rock,
      result: ResultType.Lose,
      date: dateMock,
    },
    {
      player: ChoiceName.Rock,
      computer: ChoiceName.Rock,
      result: ResultType.Tie,
      date: dateMock,
    },
  ];

  it("matches the snapshot", () => {
    const { asFragment } = render(<History history={historyMock} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the History heading and counter", () => {
    render(<History history={historyMock} />);

    expect(screen.getByText("History")).toBeInTheDocument();
    expect(screen.getByText(`${historyMock.length} / 10`)).toBeInTheDocument();
  });

  it("renders nothing if history is empty", () => {
    render(<History history={[]} />);

    expect(screen.queryAllByRole("listitem")).toHaveLength(0);
  });

  it("renders all history items correctly", () => {
    render(<History history={historyMock} />);

    const listItems = screen.getAllByRole("listitem");
    expect(listItems).toHaveLength(historyMock.length);

    // Verify the content of each list item
    listItems.forEach((item, index) => {
      const { player, computer, result } = historyMock[index];

      expect(item).toHaveTextContent(dateStringMock);
      expect(item).toHaveTextContent(`You: ${player}`);
      expect(item).toHaveTextContent(`Computer: ${computer}`);
      expect(item).toHaveTextContent(result);
    });
  });

  it("limits displayed items to 10", () => {
    const longHistory: GameResult[] = Array.from({ length: 15 }, (_, i) => ({
      player: `player${i}`,
      computer: `computer${i}`,
      result: i % 2 === 0 ? ResultType.Win : ResultType.Lose,
      date: new Date(),
    }));

    render(<History history={longHistory} />);

    expect(screen.getAllByRole("listitem")).toHaveLength(10);
  });

  it("applies correct color to result text", () => {
    render(<History history={historyMock} />);

    const results = screen.getAllByText(/win|lose|tie/i);
    results.forEach((result, index) => {
      const expectedColor = ResultToColorMap[historyMock[index].result];
      expect(result).toHaveStyle(`color: ${expectedColor}`);
    });
  });
});
