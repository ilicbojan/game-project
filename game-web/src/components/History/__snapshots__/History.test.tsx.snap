// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="history section"
  >
    <h3>
      History 
      <span>
        4 / 10
      </span>
    </h3>
    <ul>
      <li>
        <span
          class="history-date"
        >
          1/1/2025 1:00:00 AM | 
        </span>
        You chose 
        <span
          class="history-player"
        >
          rock ‚úä
        </span>
        , computer chose 
        <span
          class="history-computer"
        >
          scissors ‚úåÔ∏è
        </span>
         ‚Üí 
        <span
          class="history-result"
          style="color: green;"
        >
          win
        </span>
      </li>
      <li>
        <span
          class="history-date"
        >
          1/1/2025 1:00:00 AM | 
        </span>
        You chose 
        <span
          class="history-player"
        >
          paper ‚úã
        </span>
        , computer chose 
        <span
          class="history-computer"
        >
          spock üññ
        </span>
         ‚Üí 
        <span
          class="history-result"
          style="color: green;"
        >
          win
        </span>
      </li>
      <li>
        <span
          class="history-date"
        >
          1/1/2025 1:00:00 AM | 
        </span>
        You chose 
        <span
          class="history-player"
        >
          lizard ü¶é
        </span>
        , computer chose 
        <span
          class="history-computer"
        >
          rock ‚úä
        </span>
         ‚Üí 
        <span
          class="history-result"
          style="color: red;"
        >
          lose
        </span>
      </li>
      <li>
        <span
          class="history-date"
        >
          1/1/2025 1:00:00 AM | 
        </span>
        You chose 
        <span
          class="history-player"
        >
          rock ‚úä
        </span>
        , computer chose 
        <span
          class="history-computer"
        >
          rock ‚úä
        </span>
         ‚Üí 
        <span
          class="history-result"
          style="color: orange;"
        >
          tie
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
